libc/src/setjmp/x86_64/longjmp.cpp

```diff
@@ -7,15 +7,34 @@
 //===----------------------------------------------------------------------===//
 
 #include "src/setjmp/longjmp.h"
+#include "include/llvm-libc-macros/offsetof-macro.h"
 #include "src/__support/common.h"
 #include "src/__support/macros/config.h"
 
-#if !defined(LIBC_TARGET_ARCH_IS_X86_64)
+#if !defined(LIBC_TARGET_ARCH_IS_X86)
 #error "Invalid file include"
 #endif
 
 namespace LIBC_NAMESPACE_DECL {
 
+#ifdef __i386__
+[[noreturn]]
+LLVM_LIBC_FUNCTION(void, longjmp, (jmp_buf buf, int val)) {
+  asm(R"(
```

Comment:
I believe this asm is wrong: the input values could be referring to the registers you're writing in the asm body. You can add explicit clobber constraints for ebx/esi/edi, but I don't think you can validly clobber ebp/esp, which I believe thus means you can't use an "m" constraint at all, because the address could be an offset to ebp/esp.

You can instead pass the field offsets via "i" constraints, and a specific register that holds buf.

(I know it was already discussed and decided against, but I do want to say that I still do think it'd be safer/clearer/simpler to just use raw asm for these functions -- because being careful to ensure that you're correctly following rules of inline-asm for functions which are explicitly about breaking the typical rules (...in a controlled manner...) is _hard_!)

Reply:
> but I don't think you can validly clobber ebp/esp

You can.

```diff
diff --git a/libc/src/setjmp/x86_64/setjmp.cpp b/libc/src/setjmp/x86_64/setjmp.cpp
index 90532c5654b3..b47db5c63e18 100644
--- a/libc/src/setjmp/x86_64/setjmp.cpp
+++ b/libc/src/setjmp/x86_64/setjmp.cpp
@@ -42,7 +42,7 @@ LLVM_LIBC_FUNCTION(int, setjmp, (jmp_buf buf)) {
       [ebp] "i"(offsetof(__jmp_buf, ebp)),
       [esp] "i"(offsetof(__jmp_buf, esp)),
       [eip] "i"(offsetof(__jmp_buf, eip))
-      : "eax", "ecx");
+      : "eax", "ecx", "ebx", "esi", "edi", "ebp", "esp");
 }
 #else
 [[gnu::naked]]
```

compiles just fine.

Reply:
d'oh, that was setjmp. Indeed for longjmp, we wind up with pushes for these clobbers.

Reply:
> the input values could be referring to the registers you're writing in the asm body

Inputs are constrained via "m", not "r" or "rm".

Reply:
"m" constraint means that the operand is passed as an address, and emits the operand using syntax which specifies a memory load/store of that address (e.g. `(%rax)` vs `%rax`). It may either provide the address in a register, or with some more complex target-specific address expression (e.g. base+offset*scale on x86). But in any case, it has to use at least one register, unless the address is of a global with a load-time-constant address (not the case here), and your target supports arbitrary relocatable symbols in its memory access operands (which x86 does, but most other architectures don't).